ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/ReplLoadTask.java

```diff
@@ -0,0 +1,251 @@
+  }
+
+  /**
+   * There was a database update done before and we want to make sure we update the last repl
+   * id on this database as we are now going to switch to processing a new database.
+   */
+  private TaskTracker updateDatabaseLastReplID(int maxTasks, Context context, Scope scope)
+      throws SemanticException {
+  /*
+    we don't want to put any limits on this task as this is essential before we start
+    processing new database events.
+   */
+    TaskTracker taskTracker =
+        new AlterDatabase(context, work.databaseEvent(context.hiveConf), work.dbNameToLoadIn,
+            new TaskTracker(maxTasks)).tasks();
+    scope.rootTasks.addAll(taskTracker.tasks());
+    return taskTracker;
+  }
+
+  private void partitionsPostProcessing(BootstrapEventsIterator iterator,
+      Scope scope, TaskTracker loadTaskTracker, TaskTracker tableTracker,
+      TaskTracker partitionsTracker) throws SemanticException {
+    setUpDependencies(tableTracker, partitionsTracker);
+    if (!scope.database && !scope.table) {
+      scope.rootTasks.addAll(partitionsTracker.tasks());
+      scope.partition = true;
+    }
+    loadTaskTracker.update(tableTracker);
+    loadTaskTracker.update(partitionsTracker);
+    if (partitionsTracker.hasReplicationState()) {
+      iterator.setReplicationState(partitionsTracker.replicationState());
+    }
+  }
+
+  /*
+      This sets up dependencies such that a child task is dependant on the parent to be complete.
+   */
+  private void setUpDependencies(TaskTracker parentTasks, TaskTracker childTasks) {
+    for (Task<? extends Serializable> parentTask : parentTasks.tasks()) {
+      for (Task<? extends Serializable> childTask : childTasks.tasks()) {
+        parentTask.addDependentTask(childTask);
+      }
+    }
+  }
+
+  private void createBuilderTask(List<Task<? extends Serializable>> rootTasks,
+      boolean shouldCreateAnotherLoadTask) {
+  /*
+     do a depth first traversal here to set the dependency collection task as the final task
+     followed by the next load task so that the chain continues forward, this additional traversal
+     might not be required if we pass the dependency collection task as a context member which
+     then can be set explicitly on all tasks created further down, and we can just add the load task
+     if there is a replication state present.
+   */
+    Task<DependencyCollectionWork> dependencyTask = TaskFactory.get(new DependencyCollectionWork(), conf);
+    setupDependencyTask(rootTasks, dependencyTask);
+
+    if (shouldCreateAnotherLoadTask) {
+      Task<ReplLoadWork> loadTask = TaskFactory.get(work, conf);
+      dependencyTask.addDependentTask(loadTask);
+    }
+  }
+
+  private void setupDependencyTask(List<Task<? extends Serializable>> tasks,
+      Task<DependencyCollectionWork> dependencyTask) {
+    if (tasks != null && !tasks.isEmpty()) {
+      for (Task<? extends Serializable> task : tasks) {
+        if(task.equals(dependencyTask)) {
+          continue;
```

Comment:
it will be because when you are adding dependency task if we used the tail objects then we wouldnt reach this scenario since we are doing it for all tasks then for ex for db level you will add a child DependencyCollectionTask then we go through all of dblevel task children to add DependencyCollectionTask, the above prevents infinite recursive loop.